src/QLDQPSolver.cpp

```diff
@@ -70,41 +94,75 @@ void QLDQPSolver::updateMatrix(
 	const std::vector<GenInequality*>& genInEqConstr,
 	const std::vector<Bound*>& boundConstr)
 {
-	Aeq_.setZero();
-	Aineq_.setZero();
+	AeqFull_.setZero();
+	AineqFull_.setZero();
 	beq_.setZero();
 	bineq_.setZero();
-	XL_.fill(-std::numeric_limits<double>::infinity());
-	XU_.fill(std::numeric_limits<double>::infinity());
-	Q_.setZero();
-	C_.setZero();
+	XLFull_.fill(-std::numeric_limits<double>::infinity());
+	XUFull_.fill(std::numeric_limits<double>::infinity());
+	QFull_.setZero();
+	CFull_.setZero();
 
-	const int nrVars = int(Q_.rows());
+	const int nrVars = int(QFull_.rows());
 
 	nrAeqLines_ = 0;
-	nrAeqLines_ = fillEq(eqConstr, nrVars, nrAeqLines_, Aeq_, beq_);
+	nrAeqLines_ = fillEq(eqConstr, nrVars, nrAeqLines_, AeqFull_, beq_);
 	nrAineqLines_ = 0;
-	nrAineqLines_ = fillInEq(inEqConstr, nrVars, nrAineqLines_, Aineq_, bineq_);
-	nrAineqLines_ = fillGenInEq(genInEqConstr, nrVars, nrAineqLines_, Aineq_, bineq_);
-
-	fillBound(boundConstr, XL_, XU_);
-	fillQC(tasks, nrVars, Q_, C_);
+	nrAineqLines_ = fillInEq(inEqConstr, nrVars, nrAineqLines_, AineqFull_, bineq_);
+	nrAineqLines_ = fillGenInEq(genInEqConstr, nrVars, nrAineqLines_, AineqFull_, bineq_);
+
+	fillBound(boundConstr, XLFull_, XUFull_);
+	fillQC(tasks, nrVars, QFull_, CFull_);
+	if(dependencies_.size())
+	{
+		Aeq_.setZero();
+		Aineq_.setZero();
+		XL_.fill(-std::numeric_limits<double>::infinity());
+		XU_.fill(std::numeric_limits<double>::infinity());
+		Q_.setZero();
+		C_.setZero();
+		reduceA(AeqFull_, Aeq_, fullToReduced_, reducedToFull_, dependencies_);
+		reduceA(AineqFull_, Aineq_, fullToReduced_, reducedToFull_, dependencies_);
+		reduceBound(XLFull_, XL_, XUFull_, XU_, fullToReduced_, reducedToFull_, dependencies_);
+		reduceQC(QFull_, CFull_, Q_, C_, fullToReduced_, reducedToFull_, dependencies_);
+	}
 }
 
 
 bool QLDQPSolver::solve()
 {
-	bool success = qld_.solve(Q_, C_,
-		Aeq_.block(0, 0, nrAeqLines_, int(Aeq_.cols())), beq_.segment(0, nrAeqLines_),
-		Aineq_.block(0, 0, nrAineqLines_, int(Aineq_.cols())), bineq_.segment(0, nrAineqLines_),
-		XL_, XU_, 1e-6);
+	bool success = false;
+	if(dependencies_.size())
+	{
+		success = qld_.solve(Q_, C_,
+			Aeq_.block(0, 0, nrAeqLines_, int(Aeq_.cols())), beq_.segment(0, nrAeqLines_),
+			Aineq_.block(0, 0, nrAineqLines_, int(Aineq_.cols())), bineq_.segment(0, nrAineqLines_),
+			XL_, XU_, 1e-6);
+		expandResult(qld_.result(), XFull_,
+								 reducedToFull_,
+								 dependencies_);
+	}
+	else
+	{
+		success = qld_.solve(QFull_, CFull_,
+			AeqFull_.block(0, 0, nrAeqLines_, int(AeqFull_.cols())), beq_.segment(0, nrAeqLines_),
+			AineqFull_.block(0, 0, nrAineqLines_, int(AineqFull_.cols())), bineq_.segment(0, nrAineqLines_),
+			XLFull_, XUFull_, 1e-6);
+	}
 	return success;
 }
 
 
 const Eigen::VectorXd& QLDQPSolver::result() const
 {
-	return qld_.result();
+	if(dependencies_.size())
+	{
+		return XFull_;
+	}
+	else
+	{
+		return qld_.result();
+	}
```

Comment:
Hum you're right, I think the performance hit is probably negligible. It would be even nicer if we could make sure that dependencies.size() is a constant. I don't know if it's possible to optimize even further.